### PR TITLE
Dart plugin integration with the now implemented MOVE_FILE refactoring

### PR DESCRIPTION
**Pull request by jwren**
This change is conditional on the version of the Dart SDK being used-- the feature was enabled with version `2.3.1`

### CODE REVIEW

**Comment by alexander-doroshko** [comment]
Sorry, a bit busy today. Will review tomorrow. Meanwhile, it would be great to add a test, check corner cases and check whether `toSystemDependentPath()`/`toSystemInependentPath()` and `getConvertedOffset()`/`getOriginalOffset()` is called where needed.

File: Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java [hunk 1]
```diff
@@ -94,6 +94,8 @@
 
   public static final String MIN_SDK_VERSION = "1.12";
 
+  private static final String MIN_MOVE_FILE_SDK_VERSION = "2.3.1";
```
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java]: I'm getting 'Unable to create refactoring' with SDK 2.3.1. Maybe 2.3.2?

File: Dart/src/com/jetbrains/lang/dart/ide/refactoring/moveFile/DartServerMoveDartFileHandler.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    return dasService.isInIncludedRoots(psiFile.getVirtualFile());
+  }
+
+  @Override
+  public void prepareMovedFile(PsiFile psiFile, PsiDirectory moveDestination, Map<PsiElement, PsiElement> oldToNewMap) {
+    final Project project = psiFile.getProject();
+    final VirtualFile virtualFile = psiFile.getVirtualFile();
+
+    final String oldFilePath = FileUtil.toSystemDependentName(virtualFile.getPath());
+    final String newFilePath = FileUtil.toSystemDependentName(moveDestination.getVirtualFile().getPath()) + "/" + virtualFile.getName();
```
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/ide/refactoring/moveFile/DartServerMoveDartFileHandler.java]: Slash not converted :). 
To avoid such bugs we usually try to call `toSystemDependentName()` as late as possible. Usually, it means right before showing a string to a user, in this case, it would mean - right before sending to the server.

File: Dart/src/com/jetbrains/lang/dart/ide/refactoring/moveFile/DartServerMoveDartFileHandler.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    // Validate initial status.
+    {
+      final RefactoringStatus initialStatus = refactoring.checkInitialConditions();
+      if (initialStatus == null) {
+        return;
+      }
+
+      if (initialStatus.hasError()) {
+        final String message = initialStatus.getMessage();
+        assert message != null;
```
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/ide/refactoring/moveFile/DartServerMoveDartFileHandler.java]: Do we want to show or log this message?

File: Dart/src/com/jetbrains/lang/dart/DartBundle.properties [hunk 4]
```diff
@@ -132,6 +132,9 @@ refactoring.introduce.final.variable.dialog.title=Introduce Constant
 dart.extract.method=Extract Method
 dart.refactoring.introduce.error=Introduce Error
 dart.refactoring.introduce.anchor.error=Can't find place for the result
+dart.refactoring.move.file.dialog.title=Move File Exception
+dart.refactoring.move.file.dialog.error.applying.change=Exception when attempting to apply the move file refactoring.
+dart.refactoring.move.file.dialog.error.computing.change=Exception when attempting to compute the file changes for the file move: {{0}}"
```
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/DartBundle.properties]: `{{0}}` -> `{0}` :)

**Comment by alexander-doroshko** [comment]
I've merged the PR with slight simplifications and the mentioned fixes. A test is very much welcome in the follow-up PR. As well as any fixes that might be needed for corner cases (I didn't test manually a lot). Thanks!

**Comment by jwren** [comment]
Thanks Alex, Monday is a holiday, but I will follow up with a test next week.

WRT the versioning, yes `2.3.2` is sufficient.

**Comment by alexander-doroshko** [comment]
> I will follow up with a test

No worries, I've added a test.
